tests/common/exceptions.h

```diff
@@ -0,0 +1,302 @@
+  }
+
+  bool match(const sycl::exception& other) const {
+    if constexpr (!is_supported_by_implementation()) {
+      // There should be no compilation failures, but every CHECK_THROWS_MATCHES
+      // should fail with this matcher
+      return false;
+    } else {
+      // Error code is validated by value, with no semantic match
+      return (other.category() == expected_category()) &&
+             (other.code().value() == m_code_value);
+    }
+  }
+
+  std::string describe() const override {
+    std::ostringstream result;
+    result << "has code value " << to_integral(m_code_value);
+    result << " for backend-specific category '" << expected_category().name();
+    result << "'";
+    return result.str();
+  }
+
+ private:
+  const std::error_category& expected_category() const {
+    return sycl::error_category_for<Backend>();
+  }
+  const CodeT& m_code_value;
+};
+}  // namespace detail
+
+/**
+ *  Provides matcher for std::error_category value, with no error code
+ *  comparison
+ *
+ *  Usage example:
+ *     CHECK_THROWS_MATCHES(action, sycl::exception,
+ *         has_exception_category(sycl::sycl_category()) &&
+ *         matches_exception_code(sycl::errc::invalid));
+ */
+inline auto has_exception_category(const std::error_category& category) {
+  return detail::matcher_exception_category(category);
+}
+
+/**
+ *  Provides matcher for sycl::errc error codes with semantic match
+ *
+ *  Usage example:
+ *     CHECK_THROWS_MATCHES(action, sycl::exception,
+ *         matches_exception_code(sycl::errc::feature_not_supported) ||
+ *         matches_exception_code(sycl::errc::invalid));
+ */
+inline auto matches_exception_code(const sycl::errc& code) {
+  return detail::matcher_exception_code_semantic_match(code);
+}
+
+/**
+ *  Provides matcher for backend-specific category and error codes
+ *
+ *  Usage example:
+ *    CHECK_THROWS_MATCHES(action(), sycl::exception,
+ *        equals_exception_for<sycl::backend::opencl>(CL_INVALID_PROGRAM));
+ */
+template <sycl::backend Backend, typename CodeT>
+inline auto equals_exception_for(const CodeT& code) {
+  static_assert(std::is_same_v<CodeT, sycl::errc_for<Backend>> ||
+                std::is_same_v<CodeT, int>);
+  return detail::matcher_equals_exception_for<Backend, CodeT>(code);
+}
+
+/**
+ * Helper function to log details for SYCL exception for legacy tests
+ *
+ * Deprecated, use Catch2 macroses for new tests instead
+ */
+inline void log_exception(sycl_cts::util::logger&, const sycl::exception& e) {
+  // Print multi-line message in a single Catch2 warning
+  WARN(stringify_sycl_exception(e));
+}
+
+/**
+ * Stringification of SYCL exceptions for Catch2 tests
+ * Provides more details than simple what() message
+ *
+ * Note that neither this macro, not StringMaker specialization would provide
+ * support for expressions like `INFO("Exception: " << exception)`
+ */
+CATCH_TRANSLATE_EXCEPTION(const sycl::exception& e) {
```

Comment:
I was confused at first why this works for a translation unit that does not include `common/exceptions.h` (e.g. `id.cpp`). However it turns out this only needs to be called once from *any* translation unit, which in this case probably was `test_base.cpp`. If you still plan to add `common/exceptions.cpp` (as you mentioned in a TODO note above), I suggest to move this call there.

Reply:
I've added a `TODO` notice here as well. BTW, would it be OK to have any `.cpp` files within `tests/common` except `main.cpp`? Maybe it's better to do it within a separate PR in case of any possible architecture concerns due to `CMakeLists.txt` changes.

Reply:
I guess it should be fine, except that the name of the library (`main_funtion_object`) wouldn't be quite fitting anymore. Come to think of it, it seems to me that `stringify_sycl_exception` (as well as the call to `CATCH_TRANSLATE_EXCEPTION`) could also live in `CTS::utils`, which already consists of multiple translation units, right? At some point we need to reevaluate what goes where, but since exception printing is more of a "framework-level" responsibility, I would argue it fits there quite well.

Reply:
What do you think about moving all three headers: `conversion.h`, `enums.h` and `exceptions.h` into the `CTS::utils`?

Reply:
Sure, why not. I was just thinking, maybe we should get rid of `tests/common` altogether at some point.

Reply:
Done with renaming in a separate commit e6597c92d0561bda63d9b0172bbff9ff8b2b76e6 to avoid git history loss:

- `common/enums.h` -> `util/sycl_enums.h`
- `common/exceptions.h` -> `util/sycl_exceptions.h`
- `common/conversion.h` -> `util/conversion.h`

Also added `sycl_exceptions.cpp` file